main.tf

```
@@ -1,5 +1,48 @@
+  # 1. if var.service_account_namespace != var.service_account_name: arn:aws:iam::<account_number>:role/<namespace>-<environment>-<stage>-<optional_name>-<service_account_name>@<service_account_namespace>
+  # 2. if var.service_account_namespace == var.service_account_name: arn:aws:iam::<account_number>:role/<namespace>-<environment>-<stage>-<optional_name>-<service_account_name>
+
+  # 3. If var.service_account_namespace == "" and var.service_account_name is provided,
+  # then the role ARM will have format arn:aws:iam::<account_number>:role/<namespace>-<environment>-<stage>-<optional_name>-<service_account_name>@all,
+  # and the policy will use "StringLike" in the test condition to allow ServiceAccounts in any Kubernetes namespace to assume the role (useful for unlimited preview environments)
+
+  # 4. If var.service_account_name == "" and var.service_account_namespace is provided,
+  # then the role ARM will have format arn:aws:iam::<account_number>:role/<namespace>-<environment>-<stage>-<optional_name>-all@<service_account_namespace>,
+  # and the policy will use "StringLike" in the test condition to allow to scope the IAM role to a namespace (allow different ServiceAccounts in the same namespace to assume the role)
+  # For more details, see https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html#iam-role-configuration
+
+  # 5. If both var.service_account_name == "" and var.service_account_namespace == "",
+  # then the role ARM will have format arn:aws:iam::<account_number>:role/<namespace>-<environment>-<stage>-<optional_name>-all@all,
+  # and the policy will use "StringLike" in the test condition to allow all ServiceAccounts in all Kubernetes namespaces to assume the IAM role
+
+  service_account_namespace_provided                = var.service_account_namespace != "" && var.service_account_namespace != null
+  service_account_name_provided                     = var.service_account_name != "" && var.service_account_name != null
+  service_account_namespace_name_provided           = local.service_account_namespace_provided && local.service_account_name_provided
+  service_account_namespace_name_provided_different = local.service_account_namespace_name_provided && (var.service_account_namespace != var.service_account_name)
+
+  service_account_id_format_map = {
+    1 = format("%s@%s", var.service_account_name, var.service_account_namespace)
+    2 = var.service_account_name
+    3 = format("%s@all", var.service_account_name)
+    4 = format("all@%s", var.service_account_namespace)
+    5 = "all@all"
+  }
+
+  case = local.service_account_namespace_name_provided_different ? "1" : (
+    local.service_account_namespace_name_provided ? "2" : (
+      local.service_account_name_provided ? "3" : (
+        local.service_account_namespace_provided ? "4" : "5"
+      )
+    )
+  )
+
+  service_account_id = local.service_account_id_format_map[local.case]
```

What you have is not wrong, but it is really verbose. I prefer something compact, like:
```suggestion
  service_account_long_id = format("%v@%v", coalesce(var.service_account_name, "all"), coalesce(var.service_account_namespace, "all"))
  service_account_id = trimsuffix(local.service_account_long_id, format("@%v", var.service_account_name))
```

What do you think?

---

that's difficult to read, but clever